x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/stream_enrichment_state_machine/stream_enrichment_state_machine.ts

```diff
@@ -156,6 +165,94 @@ export const streamEnrichmentMachine = setup({
         dataSourceRef.send({ type: 'dataSource.refresh' })
       );
     },
+    clearPersistentMappings: assign(({ context }, params: { fieldName?: string }) => {
+      if (params.fieldName) {
+        const newPersistentMappings = new Map(context.persistentFieldMappings);
+        newPersistentMappings.delete(params.fieldName);
+        return { persistentFieldMappings: newPersistentMappings };
+      } else {
+        return { persistentFieldMappings: new Map() };
+      }
+    }),
+
+    storePersistentFieldMapping: assign(({ context }, params: { field: MappedSchemaField }) => {
+      try {
+        const currentFields = context.simulatorRef.getSnapshot().context.detectedSchemaFields;
+        const fieldPosition = currentFields.findIndex((f) => f.name === params.field.name);
+
+        const newPersistentMappings = new Map(context.persistentFieldMappings);
+        newPersistentMappings.set(params.field.name, {
+          fieldName: params.field.name,
+          type: params.field.type,
+          additionalParameters: params.field.additionalParameters,
+          timestamp: Date.now(),
+          lastSeenPosition: fieldPosition >= 0 ? fieldPosition : undefined,
+          processorId: undefined,
+        });
+        return { persistentFieldMappings: newPersistentMappings };
+      } catch (error) {
+        return { persistentFieldMappings: context.persistentFieldMappings };
+      }
+    }),
+    removePersistentFieldMapping: assign(({ context }, params: { fieldName: string }) => {
+      try {
+        const newPersistentMappings = new Map(context.persistentFieldMappings);
+        newPersistentMappings.delete(params.fieldName);
+        return { persistentFieldMappings: newPersistentMappings };
+      } catch (error) {
+        return { persistentFieldMappings: context.persistentFieldMappings };
+      }
+    }),
+    applyFieldRestorationFromPersistentMappings: ({ context, self }) => {
+      try {
+        const simulatorSnapshot = context.simulatorRef.getSnapshot();
+        const currentFields = simulatorSnapshot.context.detectedSchemaFields;
+
+        if (currentFields.length === 0 && context.persistentFieldMappings.size > 0) {
+          setTimeout(() => {
+            self.send({ type: 'applyFieldRestorationFromPersistentMappings' });
+          }, 1500);
+          return;
+        }
+
+        const matchingResult = matchFieldsWithPersistentMappings(
+          currentFields,
+          context.persistentFieldMappings
+        );
+
+        const significantReorderings = matchingResult.potentialReorderings.filter((reordering) => {
+          const wasRecentlyMapped = Date.now() - reordering.persistentMapping.timestamp < 5000;
```

Comment:
What about this magic number? Again, let's not make the logic so complicated - the fact that we need this shows that the solution is too complicated.